Initial commit

IIExpanse opened the pull request.

IIExpanse committed:
Initial commit

IIExpanse committed:
Initial commit

IIExpanse committed:
Initial commit

IIExpanse committed:
Initial commit

alexache commented on ewm-service/pom.xml:
```diff
@@ -0,0 +1,162 @@
+            <plugin>
+                <groupId>com.mysema.maven</groupId>
+                <artifactId>apt-maven-plugin</artifactId>
+                <version>1.1.3</version>
```
Версии зависимостей (и плагинов) должны быть указаны в родительском модуле

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/category/controller/CategoryController.java:
```diff
@@ -0,0 +1,48 @@
+    private final CategoryService service;
+
+    @PostMapping(path = "/admin/categories")
+    public ResponseEntity<CategoryDto> addCategory(@RequestBody @Valid NewCategoryDto newCategoryDto) {
```
Для чего использовать ResponseEntity, где нет необходимости в этой гибкости? Для этих контроллеров сойдет возвращение объектов ДТО

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/category/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+
+    @Override
+    public CategoryDto getCategory(long id) {
+        Optional<Category> categoryOptional = categoryRepository.findById(id);
```
Эти 7 строк можно заменить на Category category = categoryRepository.findById(id).orElseThrow(() -> new CategoryNotFoundExcveption(..)); без какой дибо потери функциональности

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/category/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+                            "Ошибка при обновлении категории: категория с id=%d не найдена.",
+                            categoryDto.getId()));
+
+        } else if (categoryRepository.existsCategoryByName(categoryDto.getName())) {
```
Что если вызвать обновление категории с тем же именем?

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/compilation/service/impl/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,154 @@
+                compilation.getEvents().stream()
+                        .map(event -> eventMapper.mapToShortDto(
+                                event,
+                                requestService.getConfirmedRequestsCount(event.getId()),
```
Это запрос к базе в цикле (в мапе стрима), такого не должно быть. Нужен один запрос по списку событий, тоже и про обращения к сервису статистики

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/model/Event.java:
```diff
@@ -0,0 +1,67 @@
+    private String title;
+    @OneToMany(mappedBy = "event")
+    @ToString.Exclude
+    Set<ParticipationRequest> requests;
```
Не хватает модификатора доступа

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/repository/EventRepository.java:
```diff
@@ -0,0 +1,62 @@
+
+    Page<Event> findAllByInitiatorId(long initiatorId, Pageable pageable);
+
+//    @Query(value = "SELECT e FROM Event e " +
```
Не должно оставаться закомментированного кода

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/event/service/impl/EventServiceImpl.java:
```diff
@@ -0,0 +1,322 @@
+                        size).stream()
+                .skip(from).map(event -> mapper.mapToFullDto(
+                        event,
+                        requestService.getConfirmedRequestsCount(event.getId()),
```
запрос к базе в цикле

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/stats/dto/ViewStats.java:
```diff
@@ -0,0 +1,17 @@
+@NoArgsConstructor
+public class ViewStats {
+
+    String app;
```
Не хватает модификаторов доступа

IIExpanse committed:
fix: 1. В родительский pom перенесены все версии зависимостей.
2. Сокращен код методов, получавших Optional из БД.
3. При обновлении категории проверка совпадении имени осуществляется только по категориям с другими id.
4. Из всех циклов убраны методы, обращающиеся к БД.
5. Добавлены отсутствовавшие модификаторы доступа.

style: Удален закомментированный код.

IIExpanse commented:
Привет, спасибо за ревью. Высылаю новую версию с учетом всех комментариев.
Из всех циклов убраны методы, обращающиеся к БД. На замену им реализованы методы, принимающие коллекции айдишников и формирующие мапы с необходимыми значениями.

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/compilation/repository/impl/CompilationRepositoryImpl.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+
+@RequiredArgsConstructor
+public class CompilationRepositoryImpl implements CompilationRepositoryCustom {
```
(опционально) действительно ли есть необходимость в таком подходе с реализацией на jdbc? Разве добавление в список событий на модели и сохранение не решит этой проблемы?

IIExpanse replied:
В логике сервиса при добавлении подборки сейчас так и делается - внутрь кладется список событий, но это ни к чему не приводит. Как я [вычитал](https://stackoverflow.com/questions/36239030/manytomany-relation-not-save), это связано с тем, что связь ManyToMany в данном случае расписана в модели события (владельце связи). 
Насколько я понимаю, для корректной работы такого сохранения надо или сохранять подборку, а потом доставать модели событий и сохранять их с подборкой внутри, либо делать владельцем связи подборку, что наверно более правильно.
Вообще я скорее потом попробую композитный ключ реализовать, чтобы ассоциативная таблица была представлена в своем собственном классе и не было таких странностей.

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/request/service/impl/RequestServiceImpl.java:
```diff
@@ -117,6 +118,17 @@ public int getConfirmedRequestsCount(long eventId) {
+        Map<Long, Integer> map = new HashMap<>();
+        ids.forEach(id -> map.put(id, 0));
+
+        requestRepository.getAllEventIdsFromConfirmedRequests(ids, RequestStatus.CONFIRMED)
```
(опционально) мне кажется, есть способ проще через подсчет количества подтвержденных заявок по событиям. Так тоже работает, но немного больше работы по обработке данных после

IIExpanse replied:
Не очень понял этот момент. Если выгружать список событий, а потом в них смотреть заявки, получится более объемно, чем простая выгрузка айдишников, как сейчас.
Я изучал возможность выгружать готовую мапу из базы, но для этого надо заморочиться с подключением еще одной библиотеки.

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/request/service/impl/RequestServiceImpl.java:
```diff
@@ -117,6 +118,17 @@ public int getConfirmedRequestsCount(long eventId) {
+        ids.forEach(id -> map.put(id, 0));
+
+        requestRepository.getAllEventIdsFromConfirmedRequests(ids, RequestStatus.CONFIRMED)
+                .forEach(id -> map.put(id, map.get(id) + 1));
```
(опционально) для такого рода изменений лучше подходит метод compute у мапы. Тогда и первоначальное наполнение нулями не потребуется

alexache commented on ewm-service/src/main/java/ru/yandex/practicum/ewm/stats/service/impl/EwmStatsServiceImpl.java:
```diff
@@ -41,4 +46,27 @@ public int getViewsForEvent(long eventId) {
+                .collect(Collectors.toList());
+
+        Map<Long, Integer> map = new HashMap<>();
+        eventIds.forEach(id -> map.put(id, 0));
```
Это операция кажется лишней

alexache commented on stats-server/src/main/java/ru/yandex/practicum/stats/service/impl/StatsServiceImpl.java:
```diff
@@ -39,22 +40,36 @@ public Collection<ViewStats> getHits(
+                String uri = hit.getUri();
+
+                if (uniqueIpsPerUriMap.get(uri).add(hit.getIp())) {
+                    hitsMap.put(uri, hitsMap.get(uri) + 1);
```
(опционально) Тут тоже лучше подойдёт compute использовать

IIExpanse commented:
Спасибо за ревью и подсказку с compute, скорректирую это и буду сливать.

IIExpanse committed:
fix: 1. Наполнение мап осуществляется через compute.

IIExpanse committed:
Revert "fix: 1. Наполнение мап осуществляется через compute."

This reverts commit a38c3fcd127e74dc6c581863f0b764c0556f07b2.

IIExpanse committed:
Откат реквеста к рабочей версии
